src/js/editor/post.js

```diff
@@ -800,18 +799,47 @@ class PostEditor {
     });
 
     let tagName = everySectionHasTagName ? 'p' : sectionTagName;
-    let firstChanged;
+    let sectionTransformations = [];
     post.walkMarkerableSections(range, section => {
       let changedSection = this.changeSectionTagName(section, tagName);
-      firstChanged = firstChanged || changedSection;
+      sectionTransformations.push({
+        from: section,
+        to: changedSection
+      });
     });
 
-    if (firstChanged) {
-      nextRange = firstChanged.headPosition().toRange();
-    }
+    let nextRange = this._determineNextRangeAfterToggleSection(range, sectionTransformations);
     this.setRange(nextRange);
   }
 
+  _determineNextRangeAfterToggleSection(range, sectionTransformations) {
+    if (sectionTransformations.length) {
+      let changedHeadSection = sectionTransformations
+        .find(({ from }) => { return from === range.headSection; })
+        .to;
```

Comment:
So this was described to me as: Walk through the changes, find the mapping which is for the same section as the old range's head, then use the mapped "to" as the new section for the head.

This seems good and like it might be a useful pattern in refactoring the range code in general.